### PR TITLE
Explore: Expand template variables when redirecting from dashboard panel

### LINKED ISSUE #19602: Explore: Add interpolateVariablesInQueries function to selected datasources
Follow up on[ PR that fixes this issue in majority of the datasources](https://github.com/grafana/grafana/pull/19582). 

When using cmd+click to open new Explore window from Dashboard, template variables in queries should be expanded. Implementation of `interpolateVariablesInQueries` function that does this in the rest if the datasources should be added to following datasources:

- [ ] Opentsdb
- [ ] Stackdriver
- [ ] Cloudwatch
- [ ] Azure monitor

### PR DESCRIPTION
**Pull request by Elfo404**
**What this PR does / why we need it**:

This is a continuation of https://github.com/grafana/grafana/pull/19582, implementing variable expansion for the following data sources:

- [X] Opentsdb
- [x] Stackdriver
- [X] Cloudwatch
- [x] Azure monitor

**Which issue(s) this PR fixes**:

Fixes #19602

**Special notes for your reviewer**:

There are some fields in some data sources that are not being interpolated as of now. I'll investigate a bit more and create another PR in case.

**EDIT:**
This https://github.com/grafana/grafana/pull/27459 will also make sure that dimension filters are going to be interpolated correctly in Azure

### CODE REVIEW

File: public/app/plugins/datasource/cloudwatch/types.ts [hunk 1]
```diff
@@ -44,6 +44,9 @@ export interface CloudWatchLogsQuery extends DataQuery {
 
 export type CloudWatchQuery = CloudWatchMetricsQuery | CloudWatchLogsQuery;
 
+export const isCloudWatchLogsQuery = (cloudwatchQuery: CloudWatchQuery): cloudwatchQuery is CloudWatchLogsQuery =>
```
Comment (aocenas) [inline public/app/plugins/datasource/cloudwatch/types.ts:47]: I think this makes sense here as it is basically a type guard.

File: public/app/plugins/datasource/opentsdb/datasource.ts [hunk 2]
```diff
@@ -493,6 +493,17 @@ export default class OpenTsDatasource extends DataSourceApi<OpenTsdbQuery, OpenT
...
   }
 
+  interpolateVariablesInQueries(queries: OpenTsdbQuery[]): OpenTsdbQuery[] {
+    if (!queries.length) {
+      return queries;
+    }
+
+    return queries.map(query => ({
+      ...query,
+      metric: this.templateSrv.replace(query.metric),
```
Comment (aocenas) [inline public/app/plugins/datasource/opentsdb/datasource.ts:503]: Probably need to have scopedVars arg here.

File: public/app/plugins/datasource/grafana-azure-monitor-datasource/datasource.ts [hunk 3]
```diff
@@ -255,4 +256,10 @@ export default class Datasource extends DataSourceApi<AzureMonitorQuery, AzureDa
   getSubscriptions() {
     return this.azureMonitorDatasource.getSubscriptions();
   }
+
+  interpolateVariablesInQueries(queries: AzureMonitorQuery[], scopedVars: ScopedVars): AzureMonitorQuery[] {
+    return queries.map(
```
Comment (zoltanbedi) [inline public/app/plugins/datasource/grafana-azure-monitor-datasource/datasource.ts:261]: You have checked the queries length in other data sources but not here. Shouldn't you check it?

File: public/app/plugins/datasource/cloud-monitoring/datasource.ts [hunk 4]
```diff
@@ -352,10 +351,14 @@ export default class CloudMonitoringDatasource extends DataSourceApi<CloudMonito
         groupBys: this.interpolateGroupBys(metricQuery.groupBys || [], scopedVars),
         view: metricQuery.view || 'FULL',
       },
-      sloQuery: this.interpolateProps(sloQuery, scopedVars),
+      sloQuery: sloQuery && this.interpolateProps(sloQuery, scopedVars),
     };
   }
 
+  interpolateVariablesInQueries(queries: CloudMonitoringQuery[], scopedVars: ScopedVars): CloudMonitoringQuery[] {
+    return queries.map(query => this.prepareTimeSeriesQuery(query, scopedVars));
```
Comment (zoltanbedi) [inline public/app/plugins/datasource/cloud-monitoring/datasource.ts:359]: You have checked the queries length in other data sources but not here. Shouldn't you check it?

**Review by zoltanbedi — APPROVED** [review]
Other than the 2 nits this looks good to me. Good job. ✨

File: public/app/plugins/datasource/grafana-azure-monitor-datasource/datasource.ts (hunk 3 above)
Comment (Elfo404) [reply public/app/plugins/datasource/grafana-azure-monitor-datasource/datasource.ts:261]: The reason is that both in this and in azure I'm only reusing pre-existing logic to interpolate variables

Comment (Elfo404) [reply public/app/plugins/datasource/grafana-azure-monitor-datasource/datasource.ts:261]: To be more specific:

while in the other datasources the fields to be interpolated are directly defined in `interpolateVariablesInQueries`, therefore the test and the implementation are coupled, here I'm using pre-existing logic that is not really tied to this function. 

Testing for how many fields get interpolated here means that whenever we change `applyTemplateVariables` in this case or `prepareTimeSeriesQuery` in cloud-monitoring we are going to break tests that are not coupled with the implementation.

I'm not really 100% happy with the tests anyway, so if you have suggestions I would really appreciate it!

File: public/app/plugins/datasource/cloud-monitoring/datasource.ts [hunk 5]
```diff
@@ -137,7 +137,8 @@ export default class CloudMonitoringDatasource extends DataSourceApi<CloudMonito
     const queries = options.targets
       .map(this.migrateQuery)
       .filter(this.shouldRunQuery)
-      .map(q => this.prepareTimeSeriesQuery(q, options));
+      .map(q => this.prepareTimeSeriesQuery(q, options.scopedVars))
+      .map(q => ({ ...q, intervalMs: options.intervalMs, type: 'timeSeriesQuery' }));
```
Comment (aocenas) [inline public/app/plugins/datasource/cloud-monitoring/datasource.ts:141]: Hmm this one new line seems like it is fixing something else than the templating isn't it? If it actually is related why is this needed?

Comment (Elfo404) [reply public/app/plugins/datasource/cloud-monitoring/datasource.ts:141]: actually I just moved it from `prepareTimeSeriesQuery` in https://github.com/grafana/grafana/pull/27354/files#diff-3c4534844b9cdbcc38d12b8239b7dfddL344 so that `prepareTimeSeriesQuery` returns a `CloudMonitoringQuery` instead of `DataQueryRequest<CloudMonitoringQuery>` which is then ugmented here. That way I can re-use it in `interpolateVariablesInQueries` without having issues with types.

Comment (aocenas) [reply public/app/plugins/datasource/cloud-monitoring/datasource.ts:141]: Ah ok did not notice that, good idea.

**Review by aocenas — APPROVED** [review]
Nice good work!

File: public/app/plugins/datasource/cloudwatch/datasource.ts [hunk 6]
```diff
@@ -958,6 +959,32 @@ export class CloudWatchDatasource extends DataSourceApi<CloudWatchQuery, CloudWa
...
+      expression: this.replace(query.expression, scopedVars),
+
+      ...(!isCloudWatchLogsQuery(query) && this.interpolateMetricsQueryVariables(query, scopedVars)),
+    }));
+  }
+
+  interpolateMetricsQueryVariables(
+    query: CloudWatchMetricsQuery,
+    scopedVars: ScopedVars
+  ): Pick<CloudWatchMetricsQuery, 'alias' | 'metricName' | 'namespace' | 'period'> {
```
Comment (papagian) [inline public/app/plugins/datasource/cloudwatch/datasource.ts:980]: I think that `dimensions` have to be interpolated too.
Consider a query like:
```
{
      "namespace": "$namespace",
      "metricName": "$metricName",
      "expression": "",
      "dimensions": {
        "InstanceId": "$instance"
      },
      "region": "$region",
      "id": "",
      "alias": "a",
      "statistics": [
        "Average"
      ],
      "period": "5m",
      "matchExact": true,
      "refId": "A"
}
```

when I navigate in explore all the other variables are interpolated except from the dimensions

**Review by papagian — CHANGES_REQUESTED** [review]
Nice work! I have tested cloud monitoring (stackdriver) and cloudwatch modifications and work like a charm; only issue I found is that cloudwatch dimensions need to be interpolated too.

File: public/app/plugins/datasource/cloudwatch/datasource.ts (hunk 6 above)
Comment (Elfo404) [reply public/app/plugins/datasource/cloudwatch/datasource.ts]: Good catch, thank you!

Should be good now :)

**Review by papagian — APPROVED** [review]
I have verified that Cloudwatch dimension interpolation works now!
I have tested also azure monitor and even though I don't have a lot of experience with it, the simple case I have tested works as expected.
